api/v1alpha1/conversion_utils.go

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ConversionAnnotationKey is the annotation key to ensure lossless
+// conversion of objects between API versions.
+const ConversionAnnotationKey = "vmoperator.vmware.com/conversion"
+
+func splitFieldsPath(input string) []string {
+	return strings.Split(input, ".")
+}
+
+// addConversionAnnotation appends the marshalled data to Object as a special
+// prefixed annotation.
+func addConversionAnnotation(obj client.Object, data map[string]interface{}) error {
```

Comment:
Minor Nit: IIRC, the idiomatic way is to return early to avoid indentation.  So maybe:
```go
func addConversionAnnotation(obj client.Object, data map[string]interface{}) error {
	if len(data) == 0 {
		return nil
	}
 	data, err := json.Marshal(data)
 	if err != nil {
 		return err
 	}
 	annotations := obj.GetAnnotations()
 	if annotations == nil {
 		annotations = map[string]string{}
 	}
 	annotations[ConversionAnnotationKey] = string(data)
 	obj.SetAnnotations(annotations)
 	return nil
 }